cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -213,9 +214,14 @@ protected Program createProgram(CConfiguration cConf, ProgramRunner programRunne
     try {
       // If the program jar is not a directory, take a snapshot of the jar file to avoid mutation.
       if (!programJarLocation.isDirectory()) {
-        programJarLocation = Locations.toLocation(Locations.linkOrCopy(programJarLocation,
-                                                                       new File(tempDir, "program.jar")));
-      }
+        File targetFile = new File(tempDir, "program.jar");
+        try {
+          programJarLocation = Locations.toLocation(Locations.linkOrCopy(programJarLocation,
+                                                                         targetFile));
+        } catch (FileAlreadyExistsException ex) {
+          LOG.warn("Program file {} already exists.", targetFile.getAbsolutePath());
```

Comment:
Should we overwrite instead? Link is atomic operation, but copy is not. Can we end up wit a partial file this way if killed in wrong moment?

Reply:
Valid point, done :)